src/ToData.purs

```diff
@@ -1,37 +1,158 @@
 module ToData
-  ( class ToData
+  ( AnotherDay(..)
+  , Day(..)
+  , Tree(..)
+  , class ConstrIndex
+  , class CountedConstrIndex
+  , class ToData
+  , class ToDataArgs
+  , class ToDataWithIndex
+  , constrIndex
+  , countedConstrIndex
+  , defaultConstrIndex
+  , genericToData
   , toData
+  , toDataArgs
+  , toDataWithIndex
   ) where
 
 import Prelude
 
 import Data.Array as Array
+import Data.Array ((..))
 import Data.BigInt (BigInt)
 import Data.BigInt as BigInt
 import Data.Either (Either(Left, Right))
 import Data.Foldable (class Foldable)
-import Data.List (List)
+import Data.Generic.Rep as G
 import Data.Map (Map)
+import Data.List (List)
 import Data.Map as Map
 import Data.Maybe (Maybe(Just, Nothing))
 import Data.Profunctor.Strong ((***))
 import Data.Ratio (Ratio, denominator, numerator)
+import Data.Symbol (class IsSymbol, SProxy(..), reflectSymbol)
+import Data.Tuple (Tuple)
 import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Helpers (uIntToBigInt)
 import Prim.TypeError (class Fail, Text)
+import Type.Proxy (Proxy(..))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Constr, Integer, List, Map, Bytes))
 
+-- Classes
+
 class ToData (a :: Type) where
   toData :: a -> PlutusData
 
+class ConstrIndex (a :: Type) where
+  constrIndex :: Proxy a -> Map String Int
+
+class ToDataWithIndex a ci where
+  toDataWithIndex :: ConstrIndex ci => Proxy ci -> a -> PlutusData
+
+-- As explained in https://harry.garrood.me/blog/write-your-own-generics/ this
+-- is just a neat pattern that flattens a skewed Product of Products
+class ToDataArgs a where
+  toDataArgs :: a -> Array (PlutusData)
+
+-- Default constructor indices
+class CountedConstrIndex (a :: Type) where
+  countedConstrIndex :: Proxy a -> Int -> Map String Int -> Map String Int
+
+-- Data.Generic.Rep instances
+
+instance toDataWithIndexSum ::
+  ( ConstrIndex a
+  , ToDataWithIndex l a
+  , ToDataWithIndex r a
+  ) =>
+  ToDataWithIndex (G.Sum l r) a where
+  toDataWithIndex p (G.Inl x) = toDataWithIndex (p :: Proxy a) x
+  toDataWithIndex p (G.Inr x) = toDataWithIndex (p :: Proxy a) x
+
+instance toDataWithIndexConstr ::
+  ( IsSymbol n
+  , ConstrIndex a
+  , ToDataArgs arg
+  ) =>
+  ToDataWithIndex (G.Constructor n arg) a where
+  toDataWithIndex p (G.Constructor args) = Constr (resolveIndex p (SProxy :: SProxy n)) (toDataArgs args)
+
+instance toDataArgsNoArguments :: ToDataArgs G.NoArguments where
+  toDataArgs _ = []
+
+instance toDataArgsArgument :: ToData a => ToDataArgs (G.Argument a) where
+  toDataArgs (G.Argument x) = [ toData x ]
+
+instance toDataArgsProduct :: (ToDataArgs a, ToDataArgs b) => ToDataArgs (G.Product a b) where
+  toDataArgs (G.Product x y) = toDataArgs x <> toDataArgs y
+
+instance (CountedConstrIndex a, CountedConstrIndex b) => CountedConstrIndex (G.Sum a b) where
+  countedConstrIndex _ i sym2ix = countedConstrIndex (Proxy :: Proxy b) (i + 1) (countedConstrIndex (Proxy :: Proxy a) i sym2ix)
```

Comment:
What's the reason behind the "flipped" ordering here? I think an example/docstring in one of my previous comments could be helpful. Since `countedConstrIndex _ i sym2ix = countedConstrIndex (Proxy :: Proxy a) i (countedConstrIndex (Proxy :: Proxy b) (i + 1) sym2ix)` also seems to compile.